nifi-nar-bundles/nifi-aws-bundle/nifi-aws-parameter-providers/src/main/java/org/apache/nifi/parameter/aws/AwsSecretsManagerParameterProvider.java

```diff
@@ -139,19 +163,35 @@ public List<ParameterGroup> fetchParameters(final ConfigurationContext context)
         AWSSecretsManager secretsManager = this.configureClient(context);
 
         final List<ParameterGroup> groups = new ArrayList<>();
-        ListSecretsRequest listSecretsRequest = new ListSecretsRequest();
-        ListSecretsResult listSecretsResult = secretsManager.listSecrets(listSecretsRequest);
-        while(!listSecretsResult.getSecretList().isEmpty()) {
-            for (final SecretListEntry entry : listSecretsResult.getSecretList()) {
-                groups.addAll(fetchSecret(secretsManager, context, entry.getName()));
+
+        // Fetch either by pattern or by enumerated list. See description of SECRET_LISTING_STRATEGY for more details.
+        String listingStrategy = context.getProperty(SECRET_LISTING_STRATEGY).getValue();
+        if (ENUMERATED_STRATEGY.equals(listingStrategy)) {
+            // if secret-name is set, fetch the secrets
+            String secretNames = context.getProperty(SECRET_NAMES).getValue();
+            for (String secretName : secretNames.split(",")) {
+                groups.addAll(fetchSecret(secretsManager, secretName));
             }
-            final String nextToken = listSecretsResult.getNextToken();
-            if (nextToken == null) {
-                break;
+        } else {
+            final Pattern secretNamePattern = Pattern.compile(context.getProperty(SECRET_NAME_PATTERN).getValue());
+            ListSecretsRequest listSecretsRequest = new ListSecretsRequest();
+            ListSecretsResult listSecretsResult = secretsManager.listSecrets(listSecretsRequest);
+            while(!listSecretsResult.getSecretList().isEmpty()) {
+                for (final SecretListEntry entry : listSecretsResult.getSecretList()) {
+                    String secretName = entry.getName();
+                    if (!secretNamePattern.matcher(secretName).matches()) {
+                        getLogger().debug("Secret [{}] does not match the secret name pattern {}", secretName, secretNamePattern);
+                        continue;
+                    }
+                    groups.addAll(fetchSecret(secretsManager, secretName));
+                }
+                final String nextToken = listSecretsResult.getNextToken();
+                if (nextToken == null) {
+                    break;
+                }
+                listSecretsRequest.setNextToken(nextToken);
+                listSecretsResult = secretsManager.listSecrets(listSecretsRequest);
```

Comment:
You could untangle determining the set of secrets to fetch from the actual fetching in here, turning one intertwined action into two separate steps.

That is, first collecting a `Set<String>` of secretNames (based on either the provided enumeration or by fetching all available secret names and then filtering for those matching the regex) and then using that set to fetch the corresponding secrets.

Reply:
Good point. Done.